// @mitter-io/core/dist/lib/MitterClientSet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var services_1 = require("./services");
var UserTokensClient_1 = require("./services/UserTokensClient");
var MitterClientSet = /** @class */ (function () {
    function MitterClientSet(mitterAxiosInterceptionHost) {
        this.mitterAxiosInterceptionHost = mitterAxiosInterceptionHost;
        this.cachedClients = {};
    }
    MitterClientSet.prototype.channels = function () {
        return this.client(services_1.ChannelsClient);
    };
    MitterClientSet.prototype.messages = function () {
        return this.client(services_1.MessagesClient);
    };
    MitterClientSet.prototype.users = function () {
        return this.client(services_1.UsersClient);
    };
    MitterClientSet.prototype.userAuth = function () {
        return this.client(UserTokensClient_1.UserTokensClient);
    };
    MitterClientSet.prototype.client = function (clientConstructor) {
        if (!(clientConstructor.name in this.cachedClients)) {
            this.cachedClients[clientConstructor.name] = new clientConstructor(this.mitterAxiosInterceptionHost);
        }
        return this.cachedClients[clientConstructor.name];
    };
    return MitterClientSet;
}());
exports.MitterClientSet = MitterClientSet;
