// @mitter-io/models/dist/lib/user/Profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.standardUserProfileAttributes = {
    firstName: 'mitter.user.profile.attribute.FirstName',
    lastName: 'mitter.user.profile.attribute.LastName',
    avatarURL: 'mitter.upa.AvatarURL',
    mobile: 'mitter.upa.Mobile',
    dob: 'mitter.upa.DOB',
    bio: 'mitter.upa.Bio',
    gender: 'mitter.upa.Gender',
    email: 'mitter.upa.Email',
    street: 'mitter.upa.address.Street',
    city: 'mitter.upa.address.City',
    state: 'mitter.upa.address.State',
    zip: 'mitter.upa.address.Zip',
    country: 'mitter.upa.address.Country'
};
var AttributeDef = /** @class */ (function () {
    function AttributeDef(key, allowedContentTypes, allowedContentEncodings, canBeEmpty, entityType // users or channels
    ) {
        this.key = key;
        this.allowedContentTypes = allowedContentTypes;
        this.allowedContentEncodings = allowedContentEncodings;
        this.canBeEmpty = canBeEmpty;
        this.entityType = entityType;
    }
    return AttributeDef;
}());
exports.AttributeDef = AttributeDef;
var Attribute = /** @class */ (function () {
    function Attribute(key, contentType, contentEncoding, value) {
        this.key = key;
        this.contentType = contentType;
        this.contentEncoding = contentEncoding;
        this.value = value;
    }
    return Attribute;
}());
var UserProfile = /** @class */ (function () {
    function UserProfile(userId, attributes) {
        this.userId = userId;
        this.attributes = attributes;
    }
    return UserProfile;
}());
exports.UserProfile = UserProfile;
