// @mitter-io/models/dist/lib/messaging/Messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StandardPayloadTypeNames;
(function (StandardPayloadTypeNames) {
    StandardPayloadTypeNames["TextMessage"] = "mitter.mt.Text";
    StandardPayloadTypeNames["FormattedTextMessage"] = "mitter.mt.FormattedText";
    StandardPayloadTypeNames["LinkInsetTextMessage"] = "mitter.mt.LinkInsetText";
    StandardPayloadTypeNames["ImageMessage"] = "mitter.mt.Image";
    StandardPayloadTypeNames["EmptyMessage"] = "mitter.mt.Empty";
    StandardPayloadTypeNames["FileMessage"] = "mitter.mt.File";
})(StandardPayloadTypeNames = exports.StandardPayloadTypeNames || (exports.StandardPayloadTypeNames = {}));
var StandardMessageType;
(function (StandardMessageType) {
    StandardMessageType["Standard"] = "Standard";
    StandardMessageType["Notification"] = "Notification";
    StandardMessageType["OutOfBand"] = "OutOfBand";
})(StandardMessageType = exports.StandardMessageType || (exports.StandardMessageType = {}));
var MessageDatum = /** @class */ (function () {
    function MessageDatum(dataType, data) {
        this.dataType = dataType;
        this.data = data;
    }
    return MessageDatum;
}());
exports.MessageDatum = MessageDatum;
var Message = /** @class */ (function () {
    function Message(senderId, textPayload, timelineEvents, messageData, appliedAcls, payloadType, messageId, messageType, entityMetaData) {
        if (messageData === void 0) { messageData = []; }
        if (payloadType === void 0) { payloadType = StandardPayloadTypeNames.TextMessage; }
        if (messageId === void 0) { messageId = null; }
        if (messageType === void 0) { messageType = StandardMessageType.Standard; }
        if (entityMetaData === void 0) { entityMetaData = { metadata: [] }; }
        this.senderId = senderId;
        this.textPayload = textPayload;
        this.timelineEvents = timelineEvents;
        this.messageData = messageData;
        this.appliedAcls = appliedAcls;
        this.payloadType = payloadType;
        this.messageId = messageId;
        this.messageType = messageType;
        this.entityMetaData = entityMetaData;
    }
    Message.prototype.identifier = function () {
        return this.messageId;
    };
    return Message;
}());
exports.Message = Message;
var ChannelReferencingMessage = /** @class */ (function () {
    function ChannelReferencingMessage(channelId, messageId, messageType, payloadType, senderId, textPayload, messageData, timelineEvents, entityMetaData) {
        if (messageType === void 0) { messageType = StandardMessageType.Standard; }
        if (payloadType === void 0) { payloadType = StandardPayloadTypeNames.TextMessage; }
        if (messageData === void 0) { messageData = []; }
        this.channelId = channelId;
        this.messageId = messageId;
        this.messageType = messageType;
        this.payloadType = payloadType;
        this.senderId = senderId;
        this.textPayload = textPayload;
        this.messageData = messageData;
        this.timelineEvents = timelineEvents;
        this.entityMetaData = entityMetaData;
    }
    ChannelReferencingMessage.prototype.identifier = function () {
        return this.messageId;
    };
    return ChannelReferencingMessage;
}());
exports.ChannelReferencingMessage = ChannelReferencingMessage;
