// @mitter-io/models/dist/lib/user/Profile.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StandardUserProfileAtrributes;
(function (StandardUserProfileAtrributes) {
    StandardUserProfileAtrributes["FirstName"] = "mitter.user.profile.attribute.FirstName";
    StandardUserProfileAtrributes["LastName"] = "mitter.user.profile.attribute.LastName";
})(StandardUserProfileAtrributes = exports.StandardUserProfileAtrributes || (exports.StandardUserProfileAtrributes = {}));
var AttributeDef = /** @class */ (function () {
    function AttributeDef(type, allowedContentTypes, allowedContentEncodings, canBeEmpty) {
        this.type = type;
        this.allowedContentTypes = allowedContentTypes;
        this.allowedContentEncodings = allowedContentEncodings;
        this.canBeEmpty = canBeEmpty;
    }
    return AttributeDef;
}());
exports.AttributeDef = AttributeDef;
var Attribute = /** @class */ (function () {
    function Attribute(type, contentType, contentEncoding, value) {
        this.type = type;
        this.contentType = contentType;
        this.contentEncoding = contentEncoding;
        this.value = value;
    }
    return Attribute;
}());
var UserProfile = /** @class */ (function () {
    function UserProfile(userId, attributes) {
        this.userId = userId;
        this.attributes = attributes;
    }
    return UserProfile;
}());
exports.UserProfile = UserProfile;
