// @mitter-io/models/dist/lib/events/central.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Central;
(function (Central) {
    Central["NewUserEvent"] = "mitter.mpe.users.NewUser";
    Central["ScreenNameChangedEvent"] = "mitter.mpe.users.ScreenNameChanged";
    Central["UserDeletedEvent"] = "mitter.mpe.users.UserDeleted";
    Central["UserLocatorAddedEvent"] = "mitter.mpe.users.UserLocatorAdded";
    Central["UserLocatorRemovedEvent"] = "mitter.mpe.users.UserLocatorRemoved";
    Central["UserTokenIssuedEvent"] = "mitter.mpe.users.UserTokenIssued";
    Central["UserTokenRevokedEvent"] = "mitter.mpe.users.UserTokenRevoked";
    Central["UserProfileAttributeDefCreatedEvent"] = "mitter.mpe.users.profiles.UserProfileAttributeDefCreatedEvent";
    Central["UserProfileModifiedEvent"] = "mitter.mpe.users.profiles.UserProfileModifiedEvent";
    Central["UserPresenceChangedEvent"] = "mitter.mpe.users.UserPresenceChanged";
    Central["NewFederatedUserEvent"] = "mitter.mpe.users.federation.NewFederatedUser";
    Central["FederatedUserLinkStateChangedEvent"] = "mitter.mpe.users.federation.FederatedUserLinkStateChanged";
    Central["SubscriberTokenIssuedEvent"] = "mitter.mpe.subscribers.SubscriberTokenIssued";
    Central["SubscriberTokenRevokedEvent"] = "mitter.mpe.subscribers.SubscriberTokenRevoked";
    Central["NewSubscriberEvent"] = "mitter.mpe.subscribers.NewSubscriber";
    Central["TokenIssuedEvent"] = "mitter.mpe.central.TokenIssuedEvent";
    Central["TokenRevokedEvent"] = "mitter.mpe.central.TokenRevokedEvent";
    Central["NewApplicationPropertyEvent"] = "mitter.mpe.applications.NewApplicationProperty";
    Central["ApplicationPropertyDeletedEvent"] = "mitter.mpe.applications.ApplicationPropertyDeleted";
    Central["ApplicationPropertyPatchedEvent"] = "mitter.mpe.applications.ApplicationPropertyPatched";
    Central["NewApplicationEvent"] = "mitter.mpe.applications.NewApplication";
    Central["NewApplicationAccessKeyEvent"] = "mitter.mpe.applications.NewApplicationAccessKey";
})(Central = exports.Central || (exports.Central = {}));
var SystemEvent = /** @class */ (function () {
    function SystemEvent() {
    }
    return SystemEvent;
}());
var ApplicationEvent = /** @class */ (function () {
    function ApplicationEvent(application) {
        this.application = application;
    }
    return ApplicationEvent;
}());
exports.ApplicationEvent = ApplicationEvent;
var SubscriberEvent = /** @class */ (function () {
    function SubscriberEvent(subscriberId) {
        this.subscriberId = subscriberId;
    }
    return SubscriberEvent;
}());
var UserEvent = /** @class */ (function (_super) {
    __extends(UserEvent, _super);
    function UserEvent(userId, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.userId = userId;
        return _this;
    }
    return UserEvent;
}(ApplicationEvent));
var NewUserEvent = /** @class */ (function (_super) {
    __extends(NewUserEvent, _super);
    function NewUserEvent(user, applicationId) {
        var _this = _super.call(this, user.userId, applicationId) || this;
        _this.user = user;
        return _this;
    }
    return NewUserEvent;
}(UserEvent));
exports.NewUserEvent = NewUserEvent;
var ScreenNameChangedEvent = /** @class */ (function (_super) {
    __extends(ScreenNameChangedEvent, _super);
    function ScreenNameChangedEvent(oldScreenName, newScreenName, userId, applicationId) {
        var _this = _super.call(this, userId, applicationId) || this;
        _this.oldScreenName = oldScreenName;
        _this.newScreenName = newScreenName;
        return _this;
    }
    return ScreenNameChangedEvent;
}(UserEvent));
exports.ScreenNameChangedEvent = ScreenNameChangedEvent;
var UserDeletedEvent = /** @class */ (function (_super) {
    __extends(UserDeletedEvent, _super);
    function UserDeletedEvent(userId, applicationId) {
        return _super.call(this, userId, applicationId) || this;
    }
    return UserDeletedEvent;
}(UserEvent));
exports.UserDeletedEvent = UserDeletedEvent;
var UserLocatorAddedEvent = /** @class */ (function (_super) {
    __extends(UserLocatorAddedEvent, _super);
    function UserLocatorAddedEvent(userLocator, userId, applicationId) {
        var _this = _super.call(this, userId, applicationId) || this;
        _this.userLocator = userLocator;
        return _this;
    }
    return UserLocatorAddedEvent;
}(UserEvent));
exports.UserLocatorAddedEvent = UserLocatorAddedEvent;
var UserLocatorRemovedEvent = /** @class */ (function (_super) {
    __extends(UserLocatorRemovedEvent, _super);
    function UserLocatorRemovedEvent(userLocator, userId, applicationId) {
        var _this = _super.call(this, userId, applicationId) || this;
        _this.userLocator = userLocator;
        return _this;
    }
    return UserLocatorRemovedEvent;
}(UserEvent));
exports.UserLocatorRemovedEvent = UserLocatorRemovedEvent;
var UserTokenRevokedEvent = /** @class */ (function (_super) {
    __extends(UserTokenRevokedEvent, _super);
    function UserTokenRevokedEvent(tokenId, userId, applicationId) {
        var _this = _super.call(this, userId, applicationId) || this;
        _this.tokenId = tokenId;
        return _this;
    }
    return UserTokenRevokedEvent;
}(UserEvent));
exports.UserTokenRevokedEvent = UserTokenRevokedEvent;
var UserPresenceChangedEvent = /** @class */ (function (_super) {
    __extends(UserPresenceChangedEvent, _super);
    function UserPresenceChangedEvent(oldPresence, newPresence, userId, applicationId) {
        var _this = _super.call(this, userId, applicationId) || this;
        _this.oldPresence = oldPresence;
        _this.newPresence = newPresence;
        return _this;
    }
    return UserPresenceChangedEvent;
}(UserEvent));
exports.UserPresenceChangedEvent = UserPresenceChangedEvent;
var UserProfileAttributeDefCreatedEvent = /** @class */ (function (_super) {
    __extends(UserProfileAttributeDefCreatedEvent, _super);
    function UserProfileAttributeDefCreatedEvent(attributeDef, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.attributeDef = attributeDef;
        return _this;
    }
    return UserProfileAttributeDefCreatedEvent;
}(ApplicationEvent));
exports.UserProfileAttributeDefCreatedEvent = UserProfileAttributeDefCreatedEvent;
var UserProfileModifiedEvent = /** @class */ (function (_super) {
    __extends(UserProfileModifiedEvent, _super);
    function UserProfileModifiedEvent(userProfile, applicationId) {
        var _this = _super.call(this, userProfile.userId, applicationId) || this;
        _this.userProfile = userProfile;
        return _this;
    }
    return UserProfileModifiedEvent;
}(UserEvent));
exports.UserProfileModifiedEvent = UserProfileModifiedEvent;
var NewApplicationPropertyEvent = /** @class */ (function (_super) {
    __extends(NewApplicationPropertyEvent, _super);
    function NewApplicationPropertyEvent(applicationProperty, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.applicationProperty = applicationProperty;
        return _this;
    }
    return NewApplicationPropertyEvent;
}(ApplicationEvent));
exports.NewApplicationPropertyEvent = NewApplicationPropertyEvent;
var ApplicationPropertyDeletedEvent = /** @class */ (function (_super) {
    __extends(ApplicationPropertyDeletedEvent, _super);
    function ApplicationPropertyDeletedEvent(applicationProperty, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.applicationProperty = applicationProperty;
        return _this;
    }
    return ApplicationPropertyDeletedEvent;
}(ApplicationEvent));
exports.ApplicationPropertyDeletedEvent = ApplicationPropertyDeletedEvent;
var ApplicationPropertyPatchedEvent = /** @class */ (function (_super) {
    __extends(ApplicationPropertyPatchedEvent, _super);
    function ApplicationPropertyPatchedEvent(applicationProperty, isDefault, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.applicationProperty = applicationProperty;
        _this.isDefault = isDefault;
        return _this;
    }
    return ApplicationPropertyPatchedEvent;
}(ApplicationEvent));
exports.ApplicationPropertyPatchedEvent = ApplicationPropertyPatchedEvent;
var NewApplicationEvent = /** @class */ (function (_super) {
    __extends(NewApplicationEvent, _super);
    function NewApplicationEvent(application, subscriberId) {
        var _this = _super.call(this, subscriberId) || this;
        _this.application = application;
        return _this;
    }
    return NewApplicationEvent;
}(SubscriberEvent));
exports.NewApplicationEvent = NewApplicationEvent;
var NewApplicationAccessKeyEvent = /** @class */ (function (_super) {
    __extends(NewApplicationAccessKeyEvent, _super);
    function NewApplicationAccessKeyEvent(accessKey, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.accessKey = accessKey;
        return _this;
    }
    return NewApplicationAccessKeyEvent;
}(ApplicationEvent));
exports.NewApplicationAccessKeyEvent = NewApplicationAccessKeyEvent;
var ApplicationAccessKeyDeletedEvent = /** @class */ (function (_super) {
    __extends(ApplicationAccessKeyDeletedEvent, _super);
    function ApplicationAccessKeyDeletedEvent(accessKey, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.accessKey = accessKey;
        return _this;
    }
    return ApplicationAccessKeyDeletedEvent;
}(ApplicationEvent));
exports.ApplicationAccessKeyDeletedEvent = ApplicationAccessKeyDeletedEvent;
var SubscriberTokenIssuedEvent = /** @class */ (function (_super) {
    __extends(SubscriberTokenIssuedEvent, _super);
    function SubscriberTokenIssuedEvent(tokenId, subscriberId) {
        var _this = _super.call(this, subscriberId) || this;
        _this.tokenId = tokenId;
        return _this;
    }
    return SubscriberTokenIssuedEvent;
}(SubscriberEvent));
exports.SubscriberTokenIssuedEvent = SubscriberTokenIssuedEvent;
var SubscriberTokenRevokedEvent = /** @class */ (function (_super) {
    __extends(SubscriberTokenRevokedEvent, _super);
    function SubscriberTokenRevokedEvent(tokenId, subscriberId) {
        var _this = _super.call(this, subscriberId) || this;
        _this.tokenId = tokenId;
        return _this;
    }
    return SubscriberTokenRevokedEvent;
}(SubscriberEvent));
exports.SubscriberTokenRevokedEvent = SubscriberTokenRevokedEvent;
var NewSubscriberEvent = /** @class */ (function (_super) {
    __extends(NewSubscriberEvent, _super);
    function NewSubscriberEvent(subscriber) {
        var _this = _super.call(this) || this;
        _this.subscriber = subscriber;
        return _this;
    }
    return NewSubscriberEvent;
}(SystemEvent));
exports.NewSubscriberEvent = NewSubscriberEvent;
var TokenIssuedEvent = /** @class */ (function (_super) {
    __extends(TokenIssuedEvent, _super);
    function TokenIssuedEvent(tokenId, entityType) {
        var _this = _super.call(this) || this;
        _this.tokenId = tokenId;
        _this.entityType = entityType;
        return _this;
    }
    return TokenIssuedEvent;
}(SystemEvent));
exports.TokenIssuedEvent = TokenIssuedEvent;
var TokenRevokedEvent = /** @class */ (function (_super) {
    __extends(TokenRevokedEvent, _super);
    function TokenRevokedEvent(tokenId, entityType) {
        var _this = _super.call(this) || this;
        _this.tokenId = tokenId;
        _this.entityType = entityType;
        return _this;
    }
    return TokenRevokedEvent;
}(SystemEvent));
exports.TokenRevokedEvent = TokenRevokedEvent;
