// @mitter-io/models/dist/lib/events/delivery-management.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var central_1 = require("./central");
var DeliveryManagementNames;
(function (DeliveryManagementNames) {
    DeliveryManagementNames["PayloadDeliveryAttemptEvent"] = "mitter.mpe.delivery.PayloadDeliveryAttempt";
    DeliveryManagementNames["UserTargetDeliveryRequestedEvent"] = "mitter.mpe.delivery.UserTargetDeliveryRequested";
    DeliveryManagementNames["DeliveryEndpointRegisteredEvent"] = "mitter.mpe.delivery.DeliveryEndpointRegistered";
    DeliveryManagementNames["DeliveryEndpointUnregisteredEvent"] = "mitter.mpe.delivery.DeliveryEndpointUnregistered";
    DeliveryManagementNames["DeliveryEndpointTransferredEvent"] = "mitter.mpe.delivery.DeliveryEndpointTransferred";
})(DeliveryManagementNames = exports.DeliveryManagementNames || (exports.DeliveryManagementNames = {}));
var PayloadDeliveryAttemptEvent = /** @class */ (function (_super) {
    __extends(PayloadDeliveryAttemptEvent, _super);
    function PayloadDeliveryAttemptEvent(serializedDeliveryEndpoint, deliveryResult, payload, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.serializedDeliveryEndpoint = serializedDeliveryEndpoint;
        _this.deliveryResult = deliveryResult;
        _this.payload = payload;
        return _this;
    }
    return PayloadDeliveryAttemptEvent;
}(central_1.ApplicationEvent));
exports.PayloadDeliveryAttemptEvent = PayloadDeliveryAttemptEvent;
var UserTargetDeliveryRequestedEvent = /** @class */ (function (_super) {
    __extends(UserTargetDeliveryRequestedEvent, _super);
    function UserTargetDeliveryRequestedEvent(user, messagingPipelinePayload, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.user = user;
        _this.messagingPipelinePayload = messagingPipelinePayload;
        return _this;
    }
    return UserTargetDeliveryRequestedEvent;
}(central_1.ApplicationEvent));
exports.UserTargetDeliveryRequestedEvent = UserTargetDeliveryRequestedEvent;
var DeliveryEndpointRegisteredEvent = /** @class */ (function (_super) {
    __extends(DeliveryEndpointRegisteredEvent, _super);
    function DeliveryEndpointRegisteredEvent(user, deliveryEndpoint, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.user = user;
        _this.deliveryEndpoint = deliveryEndpoint;
        return _this;
    }
    return DeliveryEndpointRegisteredEvent;
}(central_1.ApplicationEvent));
exports.DeliveryEndpointRegisteredEvent = DeliveryEndpointRegisteredEvent;
var DeliveryEndpointUnregisteredEvent = /** @class */ (function (_super) {
    __extends(DeliveryEndpointUnregisteredEvent, _super);
    function DeliveryEndpointUnregisteredEvent(user, deliveryEndpoint, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.user = user;
        _this.deliveryEndpoint = deliveryEndpoint;
        return _this;
    }
    return DeliveryEndpointUnregisteredEvent;
}(central_1.ApplicationEvent));
exports.DeliveryEndpointUnregisteredEvent = DeliveryEndpointUnregisteredEvent;
var DeliveryEndpointTransferredEvent = /** @class */ (function (_super) {
    __extends(DeliveryEndpointTransferredEvent, _super);
    function DeliveryEndpointTransferredEvent(deliveryEndponintId, oldOwner, newOwner, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.deliveryEndponintId = deliveryEndponintId;
        _this.oldOwner = oldOwner;
        _this.newOwner = newOwner;
        return _this;
    }
    return DeliveryEndpointTransferredEvent;
}(central_1.ApplicationEvent));
exports.DeliveryEndpointTransferredEvent = DeliveryEndpointTransferredEvent;
