// @mitter-io/models/dist/lib/events/mardle.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var central_1 = require("./central");
var ChannelEvent = /** @class */ (function (_super) {
    __extends(ChannelEvent, _super);
    function ChannelEvent(channelId, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.channelId = channelId;
        return _this;
    }
    return ChannelEvent;
}(central_1.ApplicationEvent));
var NewChannelMessageEvent = /** @class */ (function (_super) {
    __extends(NewChannelMessageEvent, _super);
    function NewChannelMessageEvent(message, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.message = message;
        return _this;
    }
    return NewChannelMessageEvent;
}(ChannelEvent));
exports.NewChannelMessageEvent = NewChannelMessageEvent;
var ChannelMessageReadyEvent = /** @class */ (function (_super) {
    __extends(ChannelMessageReadyEvent, _super);
    function ChannelMessageReadyEvent(message, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.message = message;
        return _this;
    }
    return ChannelMessageReadyEvent;
}(ChannelEvent));
exports.ChannelMessageReadyEvent = ChannelMessageReadyEvent;
var NewChannelEvent = /** @class */ (function (_super) {
    __extends(NewChannelEvent, _super);
    function NewChannelEvent(channel, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.channel = channel;
        return _this;
    }
    return NewChannelEvent;
}(ChannelEvent));
exports.NewChannelEvent = NewChannelEvent;
var ChannelMessageDeletedEvent = /** @class */ (function (_super) {
    __extends(ChannelMessageDeletedEvent, _super);
    function ChannelMessageDeletedEvent(deletedMessage, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.deletedMessage = deletedMessage;
        return _this;
    }
    return ChannelMessageDeletedEvent;
}(ChannelEvent));
exports.ChannelMessageDeletedEvent = ChannelMessageDeletedEvent;
var ChannelDeletedEvent = /** @class */ (function (_super) {
    __extends(ChannelDeletedEvent, _super);
    function ChannelDeletedEvent(deletedChannel, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.deletedChannel = deletedChannel;
        return _this;
    }
    return ChannelDeletedEvent;
}(ChannelEvent));
exports.ChannelDeletedEvent = ChannelDeletedEvent;
var ChannelParticipatedAddedEvent = /** @class */ (function (_super) {
    __extends(ChannelParticipatedAddedEvent, _super);
    function ChannelParticipatedAddedEvent(channelParticipation, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.channelParticipation = channelParticipation;
        return _this;
    }
    return ChannelParticipatedAddedEvent;
}(ChannelEvent));
exports.ChannelParticipatedAddedEvent = ChannelParticipatedAddedEvent;
var ChannelParticipationModifiedEvent = /** @class */ (function (_super) {
    __extends(ChannelParticipationModifiedEvent, _super);
    function ChannelParticipationModifiedEvent(oldChannelParticipation, newChannelParticipation, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.oldChannelParticipation = oldChannelParticipation;
        _this.newChannelParticipation = newChannelParticipation;
        return _this;
    }
    return ChannelParticipationModifiedEvent;
}(ChannelEvent));
exports.ChannelParticipationModifiedEvent = ChannelParticipationModifiedEvent;
var ChannelParticipantDeletedEvent = /** @class */ (function (_super) {
    __extends(ChannelParticipantDeletedEvent, _super);
    function ChannelParticipantDeletedEvent(channelParticipation, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.channelParticipation = channelParticipation;
        return _this;
    }
    return ChannelParticipantDeletedEvent;
}(ChannelEvent));
exports.ChannelParticipantDeletedEvent = ChannelParticipantDeletedEvent;
var NewTimelineEventForMessageEvent = /** @class */ (function (_super) {
    __extends(NewTimelineEventForMessageEvent, _super);
    function NewTimelineEventForMessageEvent(timeLineEvent, messageId, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.timeLineEvent = timeLineEvent;
        _this.messageId = messageId;
        return _this;
    }
    return NewTimelineEventForMessageEvent;
}(ChannelEvent));
exports.NewTimelineEventForMessageEvent = NewTimelineEventForMessageEvent;
var NewTimelineEventForChannelEvent = /** @class */ (function (_super) {
    __extends(NewTimelineEventForChannelEvent, _super);
    function NewTimelineEventForChannelEvent(timeLineEvent, channelId, applicationId) {
        var _this = _super.call(this, channelId, applicationId) || this;
        _this.timeLineEvent = timeLineEvent;
        return _this;
    }
    return NewTimelineEventForChannelEvent;
}(ChannelEvent));
exports.NewTimelineEventForChannelEvent = NewTimelineEventForChannelEvent;
var NewTimelineEventEvent = /** @class */ (function (_super) {
    __extends(NewTimelineEventEvent, _super);
    function NewTimelineEventEvent(timeLineEvent, applicationId) {
        var _this = _super.call(this, applicationId) || this;
        _this.timeLineEvent = timeLineEvent;
        return _this;
    }
    return NewTimelineEventEvent;
}(central_1.ApplicationEvent));
exports.NewTimelineEventEvent = NewTimelineEventEvent;
var ChannelMessagesTruncatedEvent = /** @class */ (function (_super) {
    __extends(ChannelMessagesTruncatedEvent, _super);
    function ChannelMessagesTruncatedEvent(channelId, applicationId) {
        return _super.call(this, channelId, applicationId) || this;
    }
    return ChannelMessagesTruncatedEvent;
}(ChannelEvent));
exports.ChannelMessagesTruncatedEvent = ChannelMessagesTruncatedEvent;
