// @mitter-io/models/dist/lib/wire/TimelineEventWireObjects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MessageTimelineEvent = /** @class */ (function () {
    function MessageTimelineEvent(messageId, timelineEvent, channelId) {
        this.messageId = messageId;
        this.timelineEvent = timelineEvent;
        this.channelId = channelId;
    }
    return MessageTimelineEvent;
}());
exports.MessageTimelineEvent = MessageTimelineEvent;
var MessageTimelineEventSet = /** @class */ (function () {
    function MessageTimelineEventSet(messageId, timelineEvents, channelId) {
        this.messageId = messageId;
        this.timelineEvents = timelineEvents;
        this.channelId = channelId;
    }
    return MessageTimelineEventSet;
}());
exports.MessageTimelineEventSet = MessageTimelineEventSet;
var ChannelTimelineEvent = /** @class */ (function () {
    function ChannelTimelineEvent(channelId, timelineEvent) {
        this.channelId = channelId;
        this.timelineEvent = timelineEvent;
    }
    return ChannelTimelineEvent;
}());
exports.ChannelTimelineEvent = ChannelTimelineEvent;
var ChannelTimelineEventSet = /** @class */ (function () {
    function ChannelTimelineEventSet(channelId, timelineEvents) {
        this.channelId = channelId;
        this.timelineEvents = timelineEvents;
    }
    return ChannelTimelineEventSet;
}());
exports.ChannelTimelineEventSet = ChannelTimelineEventSet;
var PatchChannelParticipation = /** @class */ (function () {
    function PatchChannelParticipation(newStatus) {
        this.newStatus = newStatus;
    }
    return PatchChannelParticipation;
}());
exports.PatchChannelParticipation = PatchChannelParticipation;
var PatchChannelParticipationResult = /** @class */ (function () {
    function PatchChannelParticipationResult(oldChannelParticipation, newChannelParticipation) {
        this.oldChannelParticipation = oldChannelParticipation;
        this.newChannelParticipation = newChannelParticipation;
    }
    return PatchChannelParticipationResult;
}());
exports.PatchChannelParticipationResult = PatchChannelParticipationResult;
