// @mitter-io/models/dist/lib/messaging/Channels.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StandardRuleSetNames;
(function (StandardRuleSetNames) {
    StandardRuleSetNames["DirectMessage"] = "io.mitter.ruleset.chats.DirectMessage";
    StandardRuleSetNames["GroupChat"] = "io.mitter.ruleset.chats.GroupChat";
    StandardRuleSetNames["SystemChannel"] = "io.mitter.ruleset.chats.SystemChannel";
    StandardRuleSetNames["SingleParticipantChannel"] = "io.mitter.ruleset.chats.SingleParticipantChannel";
})(StandardRuleSetNames = exports.StandardRuleSetNames || (exports.StandardRuleSetNames = {}));
var ParticipationStatus;
(function (ParticipationStatus) {
    ParticipationStatus["Active"] = "Active";
    ParticipationStatus["ReadOnly"] = "ReadOnly";
    ParticipationStatus["Disabled"] = "Disabled";
})(ParticipationStatus = exports.ParticipationStatus || (exports.ParticipationStatus = {}));
var Channel = /** @class */ (function () {
    function Channel(channelId, defaultRuleSet, participation, timelineEvents, systemChannel, entityMetadata, entityProfile) {
        this.channelId = channelId;
        this.defaultRuleSet = defaultRuleSet;
        this.participation = participation;
        this.timelineEvents = timelineEvents;
        this.systemChannel = systemChannel;
        this.entityMetadata = entityMetadata;
        this.entityProfile = entityProfile;
    }
    Channel.prototype.identifier = function () {
        return this.channelId;
    };
    return Channel;
}());
exports.Channel = Channel;
var ChannelParticipation = /** @class */ (function () {
    function ChannelParticipation(participantId, participationStatus, channelId) {
        if (participationStatus === void 0) { participationStatus = ParticipationStatus.Active; }
        this.participantId = participantId;
        this.participationStatus = participationStatus;
        this.channelId = channelId;
    }
    return ChannelParticipation;
}());
exports.ChannelParticipation = ChannelParticipation;
var ParticipatedChannel = /** @class */ (function () {
    function ParticipatedChannel(participationStatus, channel) {
        this.participationStatus = participationStatus;
        this.channel = channel;
    }
    return ParticipatedChannel;
}());
exports.ParticipatedChannel = ParticipatedChannel;
