// @mitter-io/models/dist/lib/delman/delivery-endpoints.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var StandardEndpointTypes;
(function (StandardEndpointTypes) {
    StandardEndpointTypes["FcmDeliveryEndpoint"] = "fcm";
    StandardEndpointTypes["TransactionalWebHook"] = "twh";
    StandardEndpointTypes["SseWebEndpoint"] = "http+sse";
})(StandardEndpointTypes || (StandardEndpointTypes = {}));
var DeliveryEndpoint = /** @class */ (function () {
    function DeliveryEndpoint(serializedEndpoint, endpointType, autoExpireAt) {
        if (autoExpireAt === void 0) { autoExpireAt = Number.MAX_SAFE_INTEGER; }
        this.serializedEndpoint = serializedEndpoint;
        this.endpointType = endpointType;
        this.autoExpireAt = autoExpireAt;
    }
    return DeliveryEndpoint;
}());
exports.DeliveryEndpoint = DeliveryEndpoint;
var FcmDeliveryEndpoint = /** @class */ (function (_super) {
    __extends(FcmDeliveryEndpoint, _super);
    function FcmDeliveryEndpoint(registrationToken) {
        var _this = _super.call(this, StandardEndpointTypes.FcmDeliveryEndpoint + ":" + registrationToken, StandardEndpointTypes.FcmDeliveryEndpoint) || this;
        _this.registrationToken = registrationToken;
        return _this;
    }
    return FcmDeliveryEndpoint;
}(DeliveryEndpoint));
exports.FcmDeliveryEndpoint = FcmDeliveryEndpoint;
var SseDeliveryEndpoint = /** @class */ (function (_super) {
    __extends(SseDeliveryEndpoint, _super);
    function SseDeliveryEndpoint(endpointToken) {
        if (endpointToken === void 0) { endpointToken = undefined; }
        var _this = _super.call(this, StandardEndpointTypes.SseWebEndpoint + ":" + endpointToken, StandardEndpointTypes.SseWebEndpoint) || this;
        _this.endpointToken = endpointToken;
        return _this;
    }
    return SseDeliveryEndpoint;
}(DeliveryEndpoint));
exports.SseDeliveryEndpoint = SseDeliveryEndpoint;
var TransactionalWebHookEndPoint = /** @class */ (function (_super) {
    __extends(TransactionalWebHookEndPoint, _super);
    function TransactionalWebHookEndPoint(webHookUri) {
        var _this = _super.call(this, StandardEndpointTypes.TransactionalWebHook + ":" + webHookUri, StandardEndpointTypes.TransactionalWebHook) || this;
        _this.webHookUri = webHookUri;
        return _this;
    }
    return TransactionalWebHookEndPoint;
}(DeliveryEndpoint));
exports.TransactionalWebHookEndPoint = TransactionalWebHookEndPoint;
var UserDeliveryEndPoint = /** @class */ (function () {
    function UserDeliveryEndPoint(userId, deliveryEndpoint, disabled) {
        if (disabled === void 0) { disabled = false; }
        this.userId = userId;
        this.deliveryEndpoint = deliveryEndpoint;
        this.disabled = disabled;
    }
    return UserDeliveryEndPoint;
}());
exports.UserDeliveryEndPoint = UserDeliveryEndPoint;
