// @mitter-io/core/dist/lib/services/UserTokensClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var common_1 = require("./common");
var constants_1 = require("./constants");
var base = constants_1.MitterConstants.Api.VersionPrefix + "/users/:userId/tokens";
exports.UserTokensPaths = {
    GetMe: base + "/me",
    GetUser: base + "/:userId",
    GetMyScreenName: base + "/me/screenname",
    GetUsersScreenName: base + "/:userIds/screenname"
};
exports.userTokensClientGenerator = common_1.clientGenerator();
var UserTokensClient = /** @class */ (function () {
    function UserTokensClient(mitterAxiosInterceptionHost) {
        this.mitterAxiosInterceptionHost = mitterAxiosInterceptionHost;
        this.userTokensAxiosClient = exports.userTokensClientGenerator(mitterAxiosInterceptionHost);
    }
    UserTokensClient.prototype.getUserToken = function (userId) {
        return this.userTokensAxiosClient
            .post("/v1/users/" + userId + "/tokens")
            .then(function (x) { return x.data; });
    };
    return UserTokensClient;
}());
exports.UserTokensClient = UserTokensClient;
