// @mitter-io/models/dist/lib/application/properties/StandardApplicationProperty.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var StandardApplicationProperty;
(function (StandardApplicationProperty) {
    StandardApplicationProperty["GoogleApiCredential"] = "google-api-credential";
    StandardApplicationProperty["GoogleOAuthCredential"] = "google-oauth-credential";
    StandardApplicationProperty["GoogleServiceAccountCredential"] = "google-service-account-credential";
    StandardApplicationProperty["AwsAccessCredentials"] = "aws-access-credentials";
    StandardApplicationProperty["ApnsProperties"] = "apns-properties";
    StandardApplicationProperty["FcmProperties"] = "fcm-properties";
    StandardApplicationProperty["OAuthIntegrationProperty"] = "oauth-integration";
    StandardApplicationProperty["TransactionalWebhookProperty"] = "transactional-webhook";
    StandardApplicationProperty["AwsSqsQueueProperty"] = "aws-sqs-queue";
    StandardApplicationProperty["AwsSnsTopicProperty"] = "aws-sns-topic";
    StandardApplicationProperty["AwsSqsEventBusProperty"] = "aws-sqs-event-bus";
    StandardApplicationProperty["AwsSnsEventBusProperty"] = "aws-sns-event-bus";
    StandardApplicationProperty["NonStandardProperty"] = "";
})(StandardApplicationProperty = exports.StandardApplicationProperty || (exports.StandardApplicationProperty = {}));
